osutil/synctree.go

```diff
@@ -0,0 +1,138 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"os"
+	"path/filepath"
+	"sort"
+	"syscall"
+)
+
+func appendWithPrefix(paths []string, prefix string, filenames []string) []string {
+	for _, filename := range filenames {
+		paths = append(paths, filepath.Join(prefix, filename))
+	}
+	return paths
+}
+
+func removeEmptyDirs(baseDir, relPath string) error {
+	for relPath != "." {
+		if err := os.Remove(filepath.Join(baseDir, relPath)); err != nil {
+			// If the directory doesn't exist, then stop.
+			if os.IsNotExist(err) {
+				return nil
+			}
+			// If the directory is not empty, then stop.
+			if pathErr, ok := err.(*os.PathError); ok && pathErr.Err == syscall.ENOTEMPTY {
+				return nil
+			}
+			return err
+		}
+		relPath = filepath.Dir(relPath)
+	}
+	return nil
+}
+
+// EnsureTreeState ensures that a directory tree content matches expectations.
+//
+// EnsureTreeState walks subdirectories of the base directory, and
+// uses EnsureDirStateGlobs to synchronise content with the
+// corresponding entry in the content map.  Any non-existent
+// subdirectories in the content map will be created.
+//
+// After synchronising all subdirectories, any subdirectories where
+// files were removed that are now empty will itself be removed, plus
+// its parent directories up to but not including the base directory.
+//
+// No checks are performed to see whether subdirectories match the
```

Comment:
It probably wouldn't be too expensive, but as I said in the comment it would probably need to be a different set of globs.  If there is a directory `snap.bar.something` and I'm trying to synchronise files matching `snap.foo.*`, the glob won't match but it is just as likely to cause problems.

Reply:
As a concrete example, imagine two invocations of the API:
```
EnsureTreeState(basedir, []string{"snap.foo.*"}, map[string]map[string]*FileState{
    "somedir/snap.bar.directory": {
        "snap.foo.filename": ...,
     },
})
EnsureTreeState(basedir, []string{"snap.bar.*"}, map[string]map[string]*FileState{
    "somedir": {
        "snap.bar.filename": ...,
     },
})
```

If we only consider the globs passed to each individual `EnsureTreeState` call, then we won't detect any problems: in both cases, the file names match the provided glob and the path components of the directories do not match the glob.  However, the first call has set a trap for the second.

Reply:
I've added sanity checks to ensure the content map is consistent with the specified globs now.  This doesn't protect against the trap described above, but I have added an extra test to ensure that an invocation like the second will error out cleanly if it encounters a problem directory.

Reply:
@jhenstridge  If I understand correctly our own usage cannot encounter the trap, right? we don't want installing a snap failing wiping out the icons of another snap.

Reply:
Right.  The icon theme PR never attempts to install files in directories matching `snap.*` (currently just ignoring such directories, so will not create this trap.